keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,969 @@
+prefix structure is `<featurename>.kubernetes.io`.
+
+<<[/UNRESOLVED]>>
+
+Policy application is controlled based on labels on the namespace. The following labels are supported:
+```
+<prefix>/allow: <policy level>
+<prefix>/allow-version: <policy version>
+<prefix>/audit: <policy level>
+<prefix>/audit-version: <policy version>
+<prefix>/warn: <policy level>
+<prefix>/warn-version: <policy version>
+```
+
+**Allow:** Pods meeting the requirements of the allowed level are allowed. Violations are rejected
+in admission.
+
+**Audit:** Pods and [templated pods] meeting the requirements of the audit policy level are ignored.
+Violations are recorded in a `<prefix>/audit: <violation>` annotation on the
+audit event for the request. Audit annotations will **not** be applied to the pod objects
+themselves, as doing so would violate the non-mutating requirement.
+
+**Warn:** Pods and [templated pods] meeting the requirements of the warn level are ignored.
+Violations are returned in a user-facing warning message. Warn & audit modes are independent; if the
+functionality of both is desired, then both labels must be set.
+
+[templated pods]: #podtemplate-resources
+
+<<[UNRESOLVED]>>
+
+The mode key names are open to discussion. In particular, I'm not satisfied that `allow` and
+`audit/warn` carry slightly different meanings:
+
+- `allow` means "allow pods meeting this profile level & below" (i.e. deny above)
+- `audit` and `warn` mean "audit/warn on pods exceeding this profile level" (i.e. audit/warn above)
+
+~~Using `deny` for the enforcing level would be more consistent with audit/warn (deny above), but is
+a counter-intuitive user experience. One would expect `deny = privileged` to deny privileged pods,
+but it really means allow privileged pods (hence using `allow` for the key).~~ (Rejected)
+
+Another option is to change the `audit` and `warn` keys to something meaning "don't audit/warn at this
+level and bellow", but we haven't thought of a good name. The best alternative we've come up with is
+`audit-above` and `warn-above`, but that is cumbersome.
+
+A final option is to change `allow` to `enforce`, which is more ambiguous. The ambiguity makes the
+inconsistency with audit & warn less obvious, but doesn't help the user experience.
+
+<<[/UNRESOLVED]>>
+
+There are several reasons for controlling the policy directly through namespace labels, rather than
+through a separate object:
+
+- Using labels enables various workflows around policy management through kubectl, for example
+  issuing queries like `kubectl get namespaces -l
+  <prefix>/allow-version!=v1.22` to find namespaces where the enforcing
+  policy isn't pinned to the most recent version.
+- Keeping the options on namespaces allows atomic create-and-set-policy, as opposed to creating a
+  namespace and then creating a second object inside the namespace.
+- Policies settings are per-namespace singletons, and singleton objects are not well supported in
+  Kubernetes.
```

Comment:
How does quota and limitranges fit? Should those get folded into namespace too?

Reply:
You can write an admission controller on "create default service account". If enforcement of those policies race with namespace creation, we don't have a privilege escalation so less is at stake.

Reply:
> How does quota and limitranges fit? Should those get folded into namespace too?

Quota objects support different scopes, so you can set pod count limits for various types of pods. You can also combine quota objects from different authors to set a backstop quota for a namespace, but give a different author control over a more fine-grained quota.